singer_sdk/testing/runners.py

```diff
@@ -0,0 +1,266 @@
+"""Utility object for running taps/targets, capturing sync output during testing."""
+
+from __future__ import annotations
+
+import abc
+import io
+import json
+from collections import defaultdict
+from contextlib import redirect_stderr, redirect_stdout
+from pathlib import Path
+from typing import IO, Any, Type, cast
+
+from singer_sdk import Tap, Target
+
+
+class SingerTestRunner(metaclass=abc.ABCMeta):
+    """Base Singer Test Runner."""
+
+    raw_messages: list[dict] = []
+    schema_messages: list[dict] = []
+    record_messages: list[dict] = []
+    state_messages: list[dict] = []
+    records: defaultdict = defaultdict(list)
+
+    def __init__(
+        self,
+        singer_class: type[Tap] | type[Target],
+        config: dict | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Initialize the test runner object.
+
+        Args:
+            singer_class (type[PluginBase]): Singer class to be tested.
+            config (dict): Tap/Target configuration for testing.
+            kwargs (dict): Default arguments to be passed to tap/target on create.
+        """
+        self.singer_class = singer_class
+        self.config = config or {}
+        self.default_kwargs = kwargs
+
+    @staticmethod
+    def _clean_sync_output(raw_records: str) -> list[dict]:
+        """Clean sync output.
+
+        Args:
+            raw_records: String containing raw messages.
+
+        Returns:
+            A list of raw messages in dict form.
+        """
+        lines = raw_records.strip().split("\n")
+        return [json.loads(ii) for ii in lines]
+
+    def create(self, kwargs: dict | None = None) -> Tap | Target:
+        """Create a new tap/target from the runner defaults.
+
+        Args:
+            kwargs (dict, optional): [description]. Defaults to None.
+
+        Returns:
+            An instantiated Tap or Target.
+        """
+        if not kwargs:
+            kwargs = self.default_kwargs
+        return self.singer_class(config=self.config, **kwargs)
```

Comment:
Should the constructor default to `parse_env_config=True`?

Reply:
No longer passed by default ✅